include/ua_plugin_history_data_backend.h

```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ *    Copyright 2018 (c) basysKom GmbH <opensource@basyskom.com> (Author: Peter Rustler)
+ */
+
+#ifndef UA_PLUGIN_HISTORY_DATA_BACKEND_H_
+#define UA_PLUGIN_HISTORY_DATA_BACKEND_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "ua_types.h"
+
+typedef struct {
+    void *context;
+    UA_StatusCode (*getHistoryData)(void * context,
```

Comment:
The plugin API should stay as close as possible to the OPC UA API.
Part 11 of the standard defines HistoryReadDetails, which can be one of 4 subtypes:
- ReadEventDetails
- ReadRawModifiedDetails
- ReadProcessedDetails
- ReadAtTimeDetails

Can we rename `getHistoryData` to `historyRead_rawModifiedDetails` and forward the ReadRawModifiedDetails structure we get from the service call? That way, we can add the other HistoryRead variations as additional callbacks later on without breaking the API.

- It is good to let the server handle continuation points and to forward a "skip" parameter to the plugin.
- What is the difference between maxSize and numValuesPerNode? Both seem to bound the max number of returned values. Since numValuesPerNode is part of the ReadRawModifiedDetails structure, maxSize seems to be redundant.
- The service defines an argument `indexRange` for the HistoryRead service. This can be parsed to `UA_NumericRange` for the internal API. Forward the `UA_NumericRange` to `historyRead_rawModifiedDetails` as well.
- The arguments should be commented to avoid misunderstandings.

Reply:
* Renaming getHistoryData to historyRead_rawModifiedDetails: I am not sure if that is a good idea. I would expect functions more low level for the plugin like insert, replace, update, delete of data. For that we should rename addHistoryData to insertHistoricalData. I would also prefer to have the heavy lifting of the answer of read raw outside of the plugin, but I have to think how to do that. For now there is no good solution. We could make getMatch part of the plugin API and move all the logic into the server outside the plugin, but I am not sure if all databases will support that way of finding values. For ReadProcessedDetails and ReadAtTimeDetails I would expect the server doing the job using getHistoryData. For ReadEventDetails I would expect an own plugin definition (not sure about that).

* Difference between maxSize and numValuesPerNode: numValuesPerNode is the count of requested values by the client. maxSize is the maximum serveable amount of values for one response by the server. So, the server will response with a continuiation point if it can not serve all by one response.

* Forward the UA_NumericRange: At the one hand this sounds good for optimizing database access, on the other hand this is only a good idea for databases that store the real types of data and do not store a bytestring of the serialized data. But if they appear in the future it will be a good idea. Agreed

* arguments commented: agreed